src/Embedders/Word2Vec.php

```diff
@@ -0,0 +1,924 @@
+
+    /**
+     * An array containing word vectors.
+     *
+     * @var Vector[]
+     */
+    protected $vectors = [];
+
+    /**
+     * The error vector used during training.
+     *
+     * @var \Tensor\Vector
+     */
+    protected $error;
+
+    /**
+     * The lock factors for each word in the context.
+     *
+     * @var int[]
+     */
+    protected $vectorsLockf = [];
+
+    /**
+     * The L2-normalized word vectors from the model.
+     *
+     * @var Vector[]
+     */
+    protected $vectorsNorm = [];
+
+    /**
+     * The cumulative distrubtion table for negative sampling.
+     *
+     * @var int[]
+     */
+    protected $cumTable = [];
+
+    /**
+     * The last digit in the cumulative distribution table.
+     *
+     * @var int
+     */
+    protected $endCumDigit;
+
+    /**
+     * The negative labels used in the cumulative distrubtion table.
+     *
+     * @var \Tensor\Vector
+     */
+    protected $negLabels;
+
+    /**
+     * The training method determined by the layer they determine.
+     *
+     * @var string
+     */
+    protected $trainMethod;
+
+    /**
+     * The layer of the network, accepts 'neg' or 'hs'.
+     *
+     * @var string
+     */
+    protected $layer;
+
+    /**
+     * The window size for the skip-gram model.
+     *
+     * @var int
+     */
+    protected $window;
+
+    /**
+     * Vector dimension size.
+     *
+     * @var int
+     */
+    protected $dimensions;
+
+    /**
+     * The degree to which noise words are removed from the training text.
+     *
+     * @var float
+     */
+    protected $sampleRate;
+
+    /**
+     * The amount of L2 regularization applied to the weights of the output layer.
+     *
+     * @var float
+     */
+    protected $alpha;
+
+    /**
+     * The maximum number of training epochs. i.e. the number of times to iterate
+     * over the entire training set before terminating.
+     *
+     * @var int
+     */
+    protected $epochs;
+
+    /**
+     * The minimum times a word must appear in the corpus to be considered in the training text.
+     *
+     * @var int
+     */
+    protected $minCount;
+
+    /**
+     * The total number of unique words in the corpus.
+     *
+     * @var int
+     */
+    protected $vocabCount;
+
+    /**
+     * Sigmoid activation function.
+     *
+     * @var \Rubix\ML\NeuralNet\ActivationFunctions\Sigmoid
+     */
+    protected $sigmoid;
+
+    /**
+     * @param string $layer
+     * @param int $window
+     * @param int $dimensions
+     * @param float $sampleRate
+     * @param float $alpha
+     * @param int $epochs
+     * @param int $minCount
+     * @throws \InvalidArgumentException
+     */
+    public function __construct(
+        string $layer = 'neg',
+        int $window = 2,
+        int $dimensions = 5,
+        float $sampleRate = 1e-3,
+        float $alpha = 0.01,
+        int $epochs = 10,
+        int $minCount = 2
+    ) {
+        if (!in_array($layer, ['neg', 'hs'])) {
```

Comment:
Did you decide to inject the training method? If so we can add a subfolder to the embedders directory/namespace to contain these objects. We'll need to think of a good name for the interface - any ideas off the top of your head?

What methods would such an interface need to provide Word2Vec?

Reply:
I need to look at it but yes, I think we should abstract the training methods. I will push this with the next release. 

As for the name of the interface, as we discussed, there really isn't one. We will have to go generic. I believe gensim struggled with this as well which is why they simply wanted boolean for which training method.

To put the differences in layman's terms, hierarchical softmax adds info about each word's Huffman encoding, negative sampling builds a cumulative distribution table for drawing random words.  

In my honest opinion, people designate these as _TrainingMethods_ and it would be most intuitive to label the interface as such.